feat : FactorisationProperties (new file)

Colin166 opened the pull request:
This file formalizes the definition of abundant, deficient, weird and semi-perfect numbers. It also proves a few theorems about them and their relationships to each other and to prime and perfect numbers.


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Colin166 committed:
Create NatNumberCategories

This file formalizes the definition of abundant, deficient, weird and semi-perfect numbers. It also proves a few theorems about them and their relationships to each other and to prime and perfect numbers.

loefflerd reviewed (commented):
Here are some comments – more to come later.

loefflerd commented on NatNumberCategories:
```diff
@@ -0,0 +1,416 @@
+    simp_all only [gt_iff_lt, le_mul_iff_one_le_right, le_mul_iff_one_le_left, mem_properDivisors,
+       true_and, ne_eq]
+
+lemma divisors_eq_proper_union_self (hn : 0 < n) :
+  Nat.divisors n = Nat.properDivisors n ∪ {n} := by
+    dsimp [divisors, properDivisors]
+    ext a; simp [*]
+    constructor
+    rintro ⟨⟨h1, h2⟩, h3⟩
+    by_cases hn : a ≠ n
+    left
+    repeat constructor
+    · assumption
+    · refine (LE.le.gt_iff_ne ?_).mpr (id (Ne.symm hn))
+      exact lt_succ.mp h2
+    · exact h3
+    repeat tauto
+    aesop
+    linarith
```
This proof is a mess. You have a non-terminal `simp`, you have an `aesop` call which fails (and it prints a message to stdout to tell you so), a whole load of `repeats` etc – this is the sort of thing which first-time Lean coders produce, throwing every conceivable tactic at the goal until something somehow sticks. 

I'm afraid I'm going to have to be brutal here: this proof shows that you do not yet have the skills needed to contribute effectively to mathlib, and that continuing with reviewing this PR is not a good use of my time. I recommend you work on your Lean coding skills, study code written by other more experienced coders, and hold off from attempting to contribute to mathlib until you are in a position to do this effectively.

FWIW, if I were trying to prove this statement, I'd probably produce something like
````
lemma divisors_eq_proper_union_self (hn : 0 < n) :
    divisors n = properDivisors n ∪ {n} := by
  ext1 a
  rw [mem_divisors, mem_union, mem_properDivisors, mem_singleton]
  constructor
  · intro ⟨hdiv, hlt⟩
    rcases eq_or_lt_of_le (le_of_dvd hn hdiv) with rfl | hlt' <;> tauto
  · rintro (⟨hdiv, _⟩ | rfl)
    · exact ⟨hdiv, hn.ne'⟩
    · exact ⟨dvd_rfl, hn.ne'⟩
````

kmill replied:
Let's judge the proof, not the contributor. Many contributors to mathlib start without knowing what makes a good contribution to mathlib, and the tough job of a reviewer is help contributors rewrite their PRs to become acceptable. Pointing out non-terminal `simp`s and unnecessary tactics is good. Providing a rewritten proof is good. But I think the other elements are nonconstructive and not something that is necessary to say.

Colin166 replied:
No worries. I am pretty new to this Mathlib stuff so I appreciate the criticism as it helps me learn. I am currently working on the file on my computer and fixing it up. When it's done I'll reopen the pull request. Thank you to @loefflerd for the suggestions and @kmill for the encouragement.

Colin166 replied:
Also I've been busy with university and research so progress has been a bit slow. But, I plan to get back into it soon.

Colin166 committed:
Update NatNumberCategories

Work in progress

Colin166 commented:
Still working on the project. Just wanted to reopen the pull request as I have made a few improvements per David's comments.

Colin166 committed:
Update FactorisationProperties

Colin166 committed:
Rename NatNumberCategories to FactorisationProperties

Colin166 commented:
I would like some help reviewing my most recent file. A few notes about the file are given.

I used "aesop" and "omega" several times throughout the file and was wondering if that is Mathlib literate. I also need help with replacing "import Mathlib" with the correct specific imports, but do not know a way to do that easily (a.k.a. not guessing and checking).

Colin166 commented:
Another issue is in "theorem seventy_weird" where I use fin_cases as a tactic. I need help with implementing that in the correct way.

loefflerd commented:
Here are a few suggestions.

- You should give your file a name ending in ".lean", and put it somewhere inside the Mathlib folder hierarchy, something like `Mathlib/NumberTheory/FactorizationProperties.lean`.
- You can get Lean to calculate a minimal set of imports for you, by putting the command `#minimize_imports` at the end of the file.
- There is a shell command described in the [instructors for contributors](https://github.com/leanprover-community/mathlib4?tab=readme-ov-file#contributing) which automatically updates the master `Mathlib.lean` file to add your new file to it. (You need to have done the first two steps already for this to work.)

Colin166 commented:
> Here are a few suggestions.
> 
> * You should give your file a name ending in ".lean", and put it somewhere inside the Mathlib folder hierarchy, something like `Mathlib/NumberTheory/FactorizationProperties.lean`.
> * You can get Lean to calculate a minimal set of imports for you, by putting the command `#minimize_imports` at the end of the file.
> * There is a shell command described in the [instructors for contributors](https://github.com/leanprover-community/mathlib4?tab=readme-ov-file#contributing) which automatically updates the master `Mathlib.lean` file to add your new file to it. (You need to have done the first two steps already for this to work.)

Thank you! Will implement.

Colin166 committed:
Rename FactorisationProperties to NumberTheory/FactorizationProperties.lean

Colin166 committed:
Update FactorizationProperties.lean

Changed file name in file
Removed unnecessary nonreference in reference section
Minimized imports

Colin166 commented:
TODO:
- I need to fix where theorem "divisors_eq_proper_union_self" and "divisor_subseteq_mul" should go in Mathlib.
- Comment above important theorems
- Anything else pointed out by reviewers

loefflerd commented:
I'm afraid the file isn't in quite the right place: you have put it in a new folder `NumberTheory/` at the top level of the git repo, whereas it should be inside the subfolder `Mathlib/`. Look to make sure you are putting it in the same place as all the other pre-existing number theory files.

Colin166 committed:
Rename NumberTheory/FactorizationProperties.lean to Mathlib/NumberTheory/FactorisationProperties

Colin166 commented:
My bad, I fixed it.

loefflerd commented:
That's the right place now, but don't forget the file name needs to have `.lean` at the end, and you need to run the update script for the top-level index file (and commit the results).

Colin166 committed:
Rename FactorisationProperties to FactorisationProperties.lean

Colin166 commented:
Okay I added the .lean to the end, but I'm not sure how to do the second thing?

Colin166 committed:
Update FactorisationProperties.lean

loefflerd committed:
Add new file to index

Colin166 commented:
Also now that I have minimized my imports, github has had trouble with several checks most notably 
continuous integration / Check all files imported (push) check. Is there a way around this?

loefflerd commented:
> Okay I added the .lean to the end, but I'm not sure how to do the second thing? [..] Also now that I have minimized my imports, github has had trouble with several checks most notably continuous integration / Check all files imported (push) check. Is there a way around this?

Github is telling you that you haven't done the second thing :-) I've pushed a commit to your branch with the necessary fix. (adding a line to the index file `Mathlib.lean` telling it to include your new file as part of Mathlib)

loefflerd committed:
Merge branch 'master' into Colin166-patch-1

loefflerd commented:
> Github is telling you that you haven't done the second thing :-) I've pushed a commit to your branch with the necessary fix. (adding a line to the index file Mathlib.lean telling it to include your new file as part of Mathlib)

Hold it, I did that wrong, sorry. Lemme just fix it a sec.

loefflerd committed:
fix

loefflerd commented:
The new file is now correctly hooked into the mathlib library; but the "lint style" checker is picking up a whole load of formatting errors, etc, which it wants you to fix, and it seems that you are also importing more files than you need to – the output of the github continuous-integration scripts will tell you what needs fixing. Over to you 🛠️

Colin166 committed:
Update FactorisationProperties.lean

Lint fixes

Colin166 commented:
Sorry was busy. Will fix the lint stuff soon.

Colin166 committed:
Update FactorisationProperties.lean

Colin166 committed:
Update FactorisationProperties.lean

Colin166 committed:
Fixed unused variables in build

Colin166 commented:
Okay lint style stuff should be fixed. However, I do have a few questions / notes before I want to admit the file.

Are aesop and omega alright to use in those cases in the file?
How do I direct Lean to put the properdivisor and divisor theorems in a different file again?

grunweg commented:
> Okay lint style stuff should be fixed. However, I do have a few questions / notes before I want to admit the file.
> 
> Are aesop and omega alright to use in those cases in the file? How do I direct Lean to put the properdivisor and divisor theorems in a different file again?

Your first question ("can I use powerful tactics or should I stick to more low-level ones") can be discussed in quite some nuance. I personally like the rule of thumb that powerful tactics are fine to use, if they actually make my life easier. If they're slow (e.g., measured subjectively), it's a good idea to look for replacements. Otherwise, you're probably fine using them.

Is your second question how to move `prop_divisors_subseteq_mul` and `divisors_eq_proper_union_self` to a different, existing file? Well, I'd say
- find an appropriate place for the lemmas. Writing `#find_home prop_divisors_subseteq_mul` (after that lemma) can be useful to suggest a good place. And your reviewer can suggest improvements to what #find_home finds.
- move each lemma to the file you want to move it to
If the new file has different `variable`s in scope, you may need to add some to your lemma - or perhaps you can remove one since the variable is already in scope.
- Make sure `FactorisationProperties.lean` still compiles: perhaps you need to `import` the file you moved your lemma to.

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+import Mathlib.Tactic.FinCases
+
+/-!
+## FactorisationProperties.lean
```
```suggestion
## Factorisation properties of natural numbers

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+
+/-!
+## FactorisationProperties.lean
+This file defines abundant, semi-perfect, deficient, and weird numbers and formalizes their
```
Follow-up question: does mathlib have perfect numbers already?
If not, the relation between Mersenne primes and even perfect numbers could be nice material to follow. Since this PR is long enough, a follow-up would be best, probably.

Colin166 replied:
Yes, I believe that would be an interesting direction to take this in the future.

Colin166 replied:
I'm also planning to add other proofs about these factorization properties later on.

Colin166 replied:
Mathlib does have perfect numbers, and I believe Aaron Anderson formalized them. Mersenne numbers were also formalized but by someone else.

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+
+## Main Definitions
+Let `n : ℕ`. All the following definitions are in the Nat namespace.
+* `Abundant` a natural number `n` is abundant if the sum of its proper divisors are greater than
```
This is a comment about conventions: I think mathlib conventions are slightly different. I would rather
- drop the line "Let `n : ℕ`. All the following definitions are in the Nat namespace." and
- qualify the lemma/definitions with `Nat`; see one suggestion below

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+* `WeirdNumber` a natural number is weird if it is both abundant and not semi-perfect
+
+## Main Results
+* `divisors_eq_proper_union_self` the `Finset` that is `Nat.divisors` is equal to the union of
+  `Nat.properDivisors` and the number itself
```
```suggestion
* `Nat.divisors_eq_proper_union_self`: `Nat.divisors n` is the union of `Nat.properDivisors` and `{n}`

grunweg replied:
One example: notable, I'd name the lemma in its namespace, and put a colon after its name.
In this case, I think the description could be more concise without losing information.

Colin166 replied:
Okay I see how to implement these suggestions, will update.

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+  http://www.jstor.org/stable/27958492. Accessed 21 Feb. 2024.
+
+## Tags
+abundant deficient weird Pseudoperfect
```
```suggestion

abundant, deficient, weird, pseudoperfect

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+
+open Nat BigOperators Finset
+
+set_option maxRecDepth 1000000
```
In general, this line makes me suspicious: is something odd going on? Was this line needed at some point and is not any more?

Colin166 replied:
Sorry I forgot to bring this up in my original comment. Yes this is necessary for proving that 70 is a weird number because fin_cases checks every iteration of the sum of each subset of the proper divisors of 70 to prove that none of them equal 70. Is there a way to centralize it only to that proof. I believe David had an issue with this initially because it could be dangerous.

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+
+variable (n m k : ℕ)
+
+/-- `n : ℕ` is Abundant if and only if the sum of the proper divisors of `n` is greater than `n`
```
I was wondering where positivity of `n` entered. What do you think about this change?
```suggestion
/-- `n : ℕ` is Abundant if and only if the sum of the proper divisors of `n` is greater than `n`.
Since `properDivisors 0 = ∅` by convention, this implies `n` is positive. -/

grunweg replied:
(And possibly making the same change to Abundant and Deficient.)

Colin166 replied:
I apologize. That was an archaic description before I had fixed the definitions of abundant and deficient to remove the 0 < n requirement

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+def Pseudoperfect (n : ℕ) : Prop :=
+  0 < n ∧ ∃ s : Finset ℕ, s ∈ powerset (properDivisors n) ∧ ∑ i in s, i = n
+
+/-- `n : ℕ` is a weird number if and only if it is both abundant and semi-perfect -/
```
This is a complete sentence: please end it with a full stop.

Colin166 replied:
I'm not sure what you mean here? I did notice that I had a typo that said semi-perfect instead of pseudoperfect. Is this what you meant?

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+      · tauto
+      · right
+        intro s
+        have hs : ¬(s ∈ powerset (properDivisors n) ∧ ∑ i in s, i = n) := by exact h2 s
```
In general, `by exact` can often be removed. Can you remove this, if it compiles?
(This also applies to many other places, I have marked a few (but not all) below.)

Colin166 replied:
Yes I've seen. Sorry, I wasn't aware of a lot of the shortcuts that lean has available, so learning this stuff is pretty cool.

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+        exact h2
+
+theorem one_deficient : Nat.Deficient 1 := by
+  dsimp [Nat.Deficient]
```
This is a so-called "non-terminal dsimp": the `dsimp` doesn't close your goal and it doesn't specify all lemmas to use.
We try to avoid these in mathlib, as adding a simp lemma could break this proof. (Unlikely in this case, but in other cases, that has happened.)

Can you replace this by an appropriate `dsimp only`? (In this case, it seems `dsimp only [Deficient]` would work; in general, you can write `dsimp?` to get a suggestion for the equivalent `dsimp only` statement.)

Colin166 replied:
Okay I did not know this. Thanks for letting me know

Colin166 replied:
I'll fix these all up

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+  norm_num
+
+theorem two_deficient : Nat.Deficient 2 := by
+  dsimp [Nat.Deficient]
```
Same here, and a few more places below.

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+      refine subset_iff.mpr ?_
+      rintro a ha
+      have ha' : a ∈ properDivisors n ∨ a ∈ properDivisors m := by
+        simp_all only [gt_iff_lt, le_mul_iff_one_le_right,
```
Can you replace this by a `simp only`? That should be a bit faster in general. (Same below; I count at least four of them in this file.)

Colin166 replied:
Yes I can. Let me fix all of those.

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+          simp_all only [mem_properDivisors, and_self]
+        refine mem_properDivisors.mpr ?_
+        constructor
+        exact Dvd.dvd.mul_right han m
```
The `constructor` call is creating two sub-goals. Please indent them appropriately with a focusing dot `·` (inserted as `\.`), e.g. `· exact Dvd.dvd.mul_right han m` and also for the `refine`.

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+          simp_all only [gt_iff_lt, le_mul_iff_one_le_right, mem_properDivisors]
+        refine mem_properDivisors.mpr ?_
+        constructor
+        exact Dvd.dvd.mul_left ham n
```
Please use a focusing dot.

Colin166 replied:
Sorry escaped my sight.

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+
+theorem prime_not_abundant (h : Prime n) : ¬ Abundant n := by
+  intro h1
+  have h2 : ∑ i in properDivisors n, i = 1 := by exact sum_properDivisors_eq_one_iff_prime.mpr h
```
Can you avoid the `by exact` here?
Does `omega` require all these hypotheses?

Colin166 replied:
I condensed the three hypotheses into two. The omega does require the first and the last ones to complete the goal.

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+  have h1 : powerset (properDivisors n) = {∅, {1}} := by
+    rw [Prime.properDivisors h]
+    exact rfl
+  have h2 : n > 1 := by exact Prime.one_lt h
```
Is this `have` actually used? I don't see where it is...

Colin166 replied:
It's used in the `linarith` a couple lines down.

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+  right
+  rintro s hs
+  have hs' : s = ∅ ∨ s = {1} := by exact List.mem_pair.mp hs
+  rcases hs' with hs1 | hs2
```
Probably, you can inline `hs'`, by writing `  rcases (List.mem_pair.mp hs) with hs1 | hs2`

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,320 @@
+  <;> {simp [*]; linarith}
+
+theorem prime_not_perfect (h : Prime n) : ¬ Perfect n := by
+  have h1 : ¬ Pseudoperfect n := by apply prime_not_pseudoperfect n h
```
Can you remove the `by apply`? If so, can you inline `h1` into the next line, i.e. write `(prime_not_pseudoperfect n h)` instead?

Colin166 replied:
Yes I can.

grunweg reviewed (commented):
I took a look and left a couple of comments about coding style. Lots of minor comments; this is perfectly normal for a first PR.

Some comments apply throughout this PR; please check the remainder of this file if you can make the same change.

Colin166 committed:
Update Mathlib/NumberTheory/FactorisationProperties.lean

Co-authored-by: grunweg <rothgami@math.hu-berlin.de>

Colin166 commented:
Thank you for your comments. I will fix the following issues soon.

Colin166 committed:
Update Mathlib/NumberTheory/FactorisationProperties.lean

Co-authored-by: grunweg <rothgami@math.hu-berlin.de>

Colin166 committed:
Update Mathlib/NumberTheory/FactorisationProperties.lean

Co-authored-by: grunweg <rothgami@math.hu-berlin.de>

Colin166 committed:
A few fixes

Removed simp_all, shortened have statements

Colin166 commented:
Okay I fixed what I could as suggested by grunweg. Here are the things that I couldn't change.

- `set_option maxRecDepth 1000000` in line 49 is used for `theorem seventy_weird` in line 111 because `fin_cases` checks every iteration to show an inequality. Is there a way to make it so that command only applies to that theorem?
- '#find_home` when placed in my file gives me the name of the file itself, not a place in Mathlib. I think the best place for the two theorems - `divisor_subseteq_mul` and `prop_divisors_subseteq_mul` - should be placed in `Mathlib.NumberTheory.Divisors.lean` as that is where `Nat.divisors` and ` Nat.properDivisors` are defined. How should I get them put there and not in my file? I don't believe any theorem in my file uses those two theorems, so removing them should keep the integrity of my file safe.
- Please check if my comments are to Mathlib standard. I believe I understood what grunweg was saying, but I will update if there is further fixing to be done.

Thanks

loefflerd commented:
Here are some more style comments.

- You still have some non-terminal simps / dsimps, despite having marked that comment of grunweg's as "resolved". Please get rid of the remaining ones.
- Please put spaces around arithmetic operations, e.g. `n ^ x` rather than `n^x`.
- I think it's best to avoid `simp [*]` or `simp only [*]` in production code. It means "simp with everything in the current context", but it's more robust, and quicker to compile, if you specify which hypotheses from the context you want to simp with.
- `:=` should be followed by one space, not two (occurs more than a dozen times).
- You often write `refine some_lemma.mpr ?_`, but that is just a complicated way of writing `rw [some_lemma]`.
- Indentation: the standard rule is that a proof of a theorem is indented 2 spaces (with further indentations within that for sub-proofs). Many of your proofs are indented more than this; I'm surprised the linter doesn't complain about this.
- Is it really necessary to have `not_pseudoperfect_not_perfect` when we already have `perfect_pseudoperfect`?
- Whenever a step in your proof generates more than one goal, you should follow it with a sequence of focussing dots to deal with the subproofs one at a time. (You do this usually but not always, e.g. in `prime_deficient`).
- If your goal is `P ∧ Q`, then rather than writing
  ````
  constructor
  · exact proof_of_P
  · exact proof_of_Q
  ````
  you can just write `exact ⟨proof_of_P, proof_of_Q⟩` (if the proofs are short enough that it fits on a line). Again I'm thinking of `prime_deficient` here, but there are other cases where this occurs in your code too. Generally, whenever you're writing `exact foo` and `foo` is fairly short, it's worth looking for a way of inlining `foo` earlier in the proof rather than letting it appear as a separate goal.
- The `aesop` call in `prime_pow_deficient` is quite slow (more than half of the running time of the entire proof). A more direct proof might be
  ````
  simp only [mem_properDivisors, mem_image, mem_range, h1]
  rintro ⟨⟨t, ht, rfl⟩, ha'⟩
  exact ⟨t, lt_of_le_of_ne ht (fun ht' ↦ lt_irrefl _ (ht' ▸ ha')), rfl⟩
  ````
  This also means you don't need to make `h2` any more.
- `linarith` can be given extra inequalities to use as arguments, so you don't need preliminary `have` statements: that is,
  ````
  have hn2 : n^x > n^y :=  (pow_lt_pow_iff_right n_gtOne).mpr hxy2
  linarith
  ````
  can be shortened to 
  ````
   linarith [(pow_lt_pow_iff_right n_gtOne).mpr hxy2]
  ```` 
   or (longer but more readable) 
  ````linarith [show n ^ x > n ^ y from (pow_lt_pow_iff_right n_gtOne).mpr hxy2]````

- Similarly, if you're using `rcases` to split up an "or" statement, you can build it inline rather than making it separately with `have`; that is,
  ````
  have hxy : x < y ∨ x > y :=  Ne.lt_or_lt hc
  rcases hxy with hxy1 | hxy2
  ````
  can simply become
  ````
  rcases Ne.lt_or_lt hc with hxy1 | hxy2
  ````
- The `<;>` tactic combinator is supposed to be used at the end of a line (not the beginning of the next). You can also write `all_goals { blah }` to apply blah to multiple goals.

loefflerd commented:
(BTW, a comment about review workflow, rather than actual code: usually the "WIP" tag is read as "ignore this, it isn't ready for other people to look at yet". So the tags "awaiting-review" and "WIP" are an odd combination.)

Colin166 commented:
Than you for your comments David. I will fix these soon.

Colin166 committed:
Update FactorisationProperties.lean

Removed non-terminal simps and replaced them with either dsimp only or simp only
Removed divisor proofs, will add to some other file as they are not necessary for this file
Made the general code more concise
Fixed formatting issues

Colin166 committed:
Update FactorisationProperties.lean

Fixed <;>

Colin166 committed:
Update FactorisationProperties.lean

Colin166 commented:
Is simp_rw alright to keep? It's in line 83.

Also, the indentations (e.g. in abundant_not_perfect_or_deficient and other theorems around it) were implemented because lint style told me to. It's because my goal line (the statement I'm trying to prove) is not in the same line as the theorem name.

loefflerd commented:
> Also, the indentations (e.g. in abundant_not_perfect_or_deficient and other theorems around it) were implemented because lint style told me to

The style linter doesn't pick up _all_ possible errors, and its suggestions for how to correct the errors it finds are extremely unreliable. If you look at the [style guide](https://leanprover-community.github.io/contribute/style.html#tactic-mode) you'll see that the indentation you're using doesn't fit the standard: the first line of the proof of any theorem should always have an indent of exactly 2 spaces (regardless of whether the theorem _statement_ fits on one line).

`simp_rw` is OK to use in library code. It is worth checking if `simp only` or plain `rw` can do what you need since they are both slightly quicker to compile, but if those don't do the job, then `simp_rw` is fine. (If you search the existing library code you'll see it's used literally thousands of times.)

I'm afraid there's still a long way to go before this code is ready for mathlib, though: the proofs are mostly still very laborious and over-long. For instance, the first theorem, `not_pseudoperfect_iff_forall`, can be condensed down to 
```lean
theorem not_pseudoperfect_iff_forall :
    ¬ Pseudoperfect n ↔ n = 0 ∨ ∀ s ∈ powerset (properDivisors n), (∑ i in s, i) ≠ n := by
  dsimp only [Pseudoperfect]
  rw [not_and_or, not_lt, nonpos_iff_eq_zero, not_exists]
  tauto
```

Colin166 commented:
Yes that's not problem. Thanks for the help. I am a novice at this, so I'm learning new stuff about theorem proving in lean every day. This pull request has helped me develop in many ways.

Colin166 commented:
I'll see what I can do to condense the code, and I'll fix the indentations. Originally, they weren't there because I did agree with the style guide: all first lines of a proof should have two spaces. I believe I accidentally mistook one of the lint style critiques as that. So, I'll fix that as well.

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,248 @@
+
+theorem perfect_pseudoperfect (h : Perfect n) : Pseudoperfect n := by
+  rcases h with ⟨h1, h2⟩
+  constructor
```
Can you golf the constructor call and the two subgoals is into one line, something like this (untested)?
```
exact ⟨h2, properDivisors n, mem_powerset_self (properDivisors n), h1⟩

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,248 @@
+theorem prime_not_perfect (h : Prime n) : ¬ Perfect n :=
+  not_pseudoperfect_not_perfect n (prime_not_pseudoperfect n h)
+
+theorem prime_deficient (h : Prime n) : Deficient n := by
```
This is a special case of `prime_pow_deficient`. Is there any particular reason you're having both?
If there is, can you at least deduce this from the general version?

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,248 @@
+  right
+  rintro s hs
+  rcases (List.mem_pair.mp hs) with hs | hs <;>
+  {simp only [ne_eq, hs, sum_empty, sum_singleton];
```
As a general comment on formatting: in general, tactic blocks using `{` and `}` can often be avoided.
As I understand it, it's preferred to you <;> twice than using a tactic block, i.e. mathlib style prefers something like
```
    rcases (List.mem_pair.mp hs) with hs | hs <;>
    simp only [ne_eq, hs, sum_empty, sum_singleton]<;>
    exact Nat.ne_of_lt (by linarith [Prime.one_lt h])}

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,248 @@
+theorem exists_infinite_even_deficients : ∃ d, n ≤ d ∧ Deficient d ∧ Even d := by
+  use 2 ^ (n + 1)
+  constructor
+  · rw [Nat.le_iff_lt_or_eq]
```
Does mathlib have this estimate already? If not, please extract this into a separate lemma, for any base `a : N`.

loefflerd replied:
There's `Nat.lt_two_pow` which is considerably stronger, it shows `n < 2 ^ n`. This is actually used in the proof here, although in a strangely roundabout way.

grunweg commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,248 @@
+
+theorem exists_infinite_odd_deficients : ∃ d, n ≤ d ∧ Nat.Deficient d ∧ Odd d := by
+  use 3 ^ (n + 1)
+  have nlttwo : n < 2 ^ (n + 1) := by
```
Then, you can use the above lemma here.

grunweg reviewed (commented):
I have a few more style comments. This list is not exhaustive; I hope these are helpful still.

loefflerd commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,248 @@
+  · exact ⟨prime_not_abundant n h, prime_not_perfect n h⟩
+  · exact Prime.pos h
+
+/-There exists infinite deficient numbers-/
```
Possibly not – particularly since there's also `exists_infinite_even_deficients` and `exists_infinite_odd_deficients`.

loefflerd replied:
(Incidentally, docstrings need to begin with `/--` (two dashes) rather than just `/-` or they don't get picked up by the reference-manual-builder script; and the start/end delimiters should each be separated from the text with a space.)

Colin166 commented:
Sorry was busy with school and work. I'm back again and will fix these issues soon.

Colin166 commented:
Thanks for your comments by the way. They help me develop as a Lean 4 theorem prover. I did not know you could fold the exists statement into an exact tactic

Colin166 commented on Mathlib/NumberTheory/FactorisationProperties.lean:
```diff
@@ -0,0 +1,248 @@
+* `Nat.deficient_or_perfect_or_abundant`: for 0 < `n:ℕ` it is either deficient, perfect, or
+  abundant
+* `Nat.prime_deficient`: all prime natural numbers are deficient
+* `Nat.exists_infinite_deficients`: there are infinite deficient numbers
```
Haha true, I'll fix this.

Colin166 committed:
Update FactorisationProperties.lean

Colin166 committed:
Update FactorisationProperties.lean

Colin166 committed:
Update FactorisationProperties.lean

Colin166 commented:
Do you guys know how to fix the build error I'm getting? This hasn't happened before, and I haven't changed anything drastically with the file.

grunweg commented:
Can you merge `master`? I believe this particular error is usually fixed by that.

Colin166 commented:
How do I do that?

loefflerd commented:
Just open a command prompt, go to the top-level directory of your git checkout (the one containing the file "Mathlib.lean"), and type "git fetch" followed by "git merge origin/master".

Colin166 committed:
Update Mathlib.lean

Colin166 committed:
Update Mathlib.lean

Colin166 committed:
Update Mathlib.lean

Colin166 commented:
I'm going to close this pull request and open a new one with my updated file.
